src/util/file-filter.js

```diff
@@ -0,0 +1,110 @@
+/* @flow */
+import path from 'path';
+
+import minimatch from 'minimatch';
+
+import {createLogger} from './logger';
+
+const log = createLogger(__filename);
+
+// Use this function to mimic path.resolve without resolving to absolute path.
+export const normalizeResolve = (file: string): string => {
+  // normalize
+  file = path.normalize(file);
+
+  // trim trailing slash
+  if (path.parse(file).base && file.endsWith(path.sep)) {
+    return file.slice(0, -1);
+  }
+  return file;
+};
+
+// FileFilter types and implementation.
+
+export type FileFilterOptions = {|
+  filesToIgnore?: Array<string>,
+  ignoreFiles?: Array<string>,
+  sourceDir?: string,
+  artifactsDir?: string,
+|};
+
+/*
+ * Allows or ignores files.
+ */
+export class FileFilter {
+  filesToIgnore: Array<string>;
+  sourceDir: string;
+
+  constructor({
+    filesToIgnore = [
+      '**/*.xpi',
+      '**/*.zip',
+      '**/.*', // any hidden file and folder
+      '**/.*/**/*', // and the content inside hidden folder
+      '**/node_modules',
+      '**/node_modules/**/*',
+    ],
+    ignoreFiles = [],
+    sourceDir = '',
```

Comment:
`sourceDir?: string;` cause parsing error in eslint. Is `sourceDir: string | typeof undefined;` correct?